kolibri/core/logger/upgrade.py

```diff
@@ -13,3 +16,31 @@ def migrate_examlog_and_examattemptlogs():
     data structures.
     """
     migrate_from_exam_logs(ExamLog.objects.all())
+
+
+@version_upgrade(old_version="<0.15.11")
+def fix_asymptotic_progress():
+    """
+    Migrate any ContentSummaryLogs that have erroneously had their progress set to >0.999 but not to 1
+    """
+    for summarylog in ContentSummaryLog.objects.filter(
+        progress__gt=0.999, progress__lt=1
+    ):
+        # Do this in a for loop rather than trying to minimize write queries, as it's a one time event
+        # and the queries for finding any correlated mastery log would be quite complex joins.
+        summarylog.progress = 1
+        summarylog.completion_timestamp = summarylog.end_timestamp
```

Comment:
I guess technically it should/could look at the attempts or sessions to determine a possibly earlier actual completion time, but that might get expensive, yeah.